src/Implement/DownloadStream.cpp

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2021, 鍾淯丞, 周杰仕, 林仁鴻. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "DownloadStream.h"
+
+#include <fstream>
+#include <string>
+
+DownloadStream::DownloadStream(const std::string &filename,
+                               const unsigned long &chunk_size)
+    : filename(filename), chunk_size(chunk_size) {
+  fd.open(filename, std::fstream::in);
+  reading = true;
+  if (!fd) {
+    reading = false;
+  }
+}
+
+int DownloadStream::ReadChunk(unsigned long &pos, std::string *const binary) {
+  if (!reading) {
+    return 0;
+  }
+
+  fd.seekg(pos * chunk_size, std::ios::beg);
+  char buffer[chunk_size + 1] = "";
+  fd.read(buffer, chunk_size);
+  *binary = buffer;
```

Comment:
string(const char*) construct the string with NULL terminated strings. This might not work if the file contains '\0'.

Reply:
Assuming the file is plaintext, there won't be '\0' in the middle of file content.